File: src/desktop/modern/domain/models/arrow_data.py
"""
Pictograph Domain Models for Kinetic Constructor

These models represent the pure business logic for pictograph rendering and positioning.
They integrate with the existing modern architecture and eliminate UI coupling from the
original pictograph implementation.

REPLACES:
- base_widgets.pictograph.pictograph.Pictograph (QGraphicsScene)
- base_widgets.pictograph.elements.pictograph_elements.PictographElements
- base_widgets.pictograph.state.pictograph_state.PictographState
- Complex UI-coupled pictograph classes

PROVIDES:
- Immutable pictograph data structures
- Pure business logic for arrow and prop positioning
- Clean separation between data and rendering
- Easy testing and serialization
"""

from __future__ import annotations

from dataclasses import dataclass, field
import json
from typing import Any
import uuid

from desktop.modern.domain.models.enums import ArrowType


@dataclass(frozen=True)
class ArrowData:
    """
    Immutable data for an arrow in a pictograph.

    REPLACES: objects.arrow.arrow.Arrow (with UI coupling)
    """

    # Core identity
    id: str = field(default_factory=lambda: str(uuid.uuid4()))
    arrow_type: ArrowType = ArrowType.BLUE

    # Visual properties
    color: str = "blue"
    turns: float = 0.0
    is_mirrored: bool = False

    # Position data (calculated by positioning system)
    location: str | None = None
    position_x: float = 0.0
    position_y: float = 0.0
    rotation_angle: float = 0.0

    # State flags
    is_visible: bool = True
    is_selected: bool = False

    def to_dict(self) -> dict[str, Any]:
        """Convert to dictionary for serialization."""
        # Handle both enum and string values for arrow_type
        arrow_type_value = (
            self.arrow_type.value
            if hasattr(self.arrow_type, "value")
            else self.arrow_type
        )

        # Note: motion_data is deprecated - motion data now lives in PictographData.motions
        return {
            "id": self.id,
            "arrow_type": arrow_type_value,
            "color": self.color,
            "turns": self.turns,
            "is_mirrored": self.is_mirrored,
            "location": self.location,
            "position_x": self.position_x,
            "position_y": self.position_y,
            "rotation_angle": self.rotation_angle,
            "is_visible": self.is_visible,
            "is_selected": self.is_selected,
        }

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> ArrowData:
        """Create from dictionary."""
        # Note: motion_data is deprecated - motion data now lives in PictographData.motions

        return cls(
            id=data.get("id", str(uuid.uuid4())),
            arrow_type=ArrowType(data.get("arrow_type", "blue")),
            color=data.get("color", "blue"),
            turns=data.get("turns", 0.0),
            is_mirrored=data.get("is_mirrored", False),
            location=data.get("location"),
            position_x=data.get("position_x", 0.0),
            position_y=data.get("position_y", 0.0),
            rotation_angle=data.get("rotation_angle", 0.0),
            is_visible=data.get("is_visible", True),
            is_selected=data.get("is_selected", False),
        )

    def to_camel_dict(self) -> dict[str, Any]:
        """Convert to dictionary with camelCase keys for JSON APIs."""
        from ..serialization import dataclass_to_camel_dict

        return dataclass_to_camel_dict(self)

    def to_json(self, camel_case: bool = True, **kwargs) -> str:
        """Serialize to JSON string."""
        from ..serialization import domain_model_to_json

        if camel_case:
            return domain_model_to_json(self, **kwargs)
        return json.dumps(self.to_dict(), **kwargs)

    @classmethod
    def from_json(cls, json_str: str, camel_case: bool = True) -> ArrowData:
        """Create instance from JSON string."""
        from ..serialization import domain_model_from_json

        if camel_case:
            return domain_model_from_json(json_str, cls)
        data = json.loads(json_str)
        return cls.from_dict(data)

File: src/web_app/modern_web/src/lib/domain/ArrowData.ts
/**
 * Arrow Domain Model
 *
 * Immutable data for an arrow in a pictograph.
 * Based on modern desktop app's arrow_data.py
 */

import { ArrowType } from './enums';

export interface ArrowData {
	readonly id: string;
	readonly arrow_type: ArrowType;
	readonly color: string;
	readonly turns: number;
	readonly is_mirrored: boolean;

	// Motion properties
	readonly motion_type: string;
	readonly start_orientation: string;
	readonly end_orientation: string;
	readonly rotation_direction: string;

	// Position data (calculated by positioning system)
	readonly location?: string | null;
	readonly position_x: number;
	readonly position_y: number;
	readonly rotation_angle: number;
	readonly coordinates?: { x: number; y: number } | null;
	readonly svg_center?: any;
	readonly svg_mirrored?: boolean;

	// State flags
	readonly is_visible: boolean;
	readonly is_selected: boolean;
}

export function createArrowData(data: Partial<ArrowData> = {}): ArrowData {
	return {
		id: data.id ?? crypto.randomUUID(),
		arrow_type: data.arrow_type ?? ArrowType.BLUE,
		color: data.color ?? 'blue',
		turns: data.turns ?? 0.0,
		is_mirrored: data.is_mirrored ?? false,
		motion_type: data.motion_type ?? 'static',
		start_orientation: data.start_orientation ?? 'in',
		end_orientation: data.end_orientation ?? 'in',
		rotation_direction: data.rotation_direction ?? 'clockwise',
		location: data.location ?? null,
		position_x: data.position_x ?? 0.0,
		position_y: data.position_y ?? 0.0,
		rotation_angle: data.rotation_angle ?? 0.0,
		coordinates: data.coordinates ?? null,
		svg_center: data.svg_center ?? null,
		svg_mirrored: data.svg_mirrored ?? false,
		is_visible: data.is_visible ?? true,
		is_selected: data.is_selected ?? false,
	};
}

export function updateArrowData(arrow: ArrowData, updates: Partial<ArrowData>): ArrowData {
	return {
		...arrow,
		...updates,
	};
}

export function arrowDataToObject(arrow: ArrowData): Record<string, any> {
	return {
		id: arrow.id,
		arrow_type: arrow.arrow_type,
		color: arrow.color,
		turns: arrow.turns,
		is_mirrored: arrow.is_mirrored,
		location: arrow.location,
		position_x: arrow.position_x,
		position_y: arrow.position_y,
		rotation_angle: arrow.rotation_angle,
		is_visible: arrow.is_visible,
		is_selected: arrow.is_selected,
	};
}

export function arrowDataFromObject(data: Record<string, any>): ArrowData {
	return createArrowData({
		id: data.id,
		arrow_type: data.arrow_type,
		color: data.color,
		turns: data.turns,
		is_mirrored: data.is_mirrored,
		location: data.location,
		position_x: data.position_x,
		position_y: data.position_y,
		rotation_angle: data.rotation_angle,
		is_visible: data.is_visible,
		is_selected: data.is_selected,
	});
}
